clang/lib/CConv/ProgramInfo.cpp

```diff
@@ -404,10 +413,31 @@ ProgramInfo::insertIntoExternalFunctionMap(ExternalFunctionMapType &Map,
         newC->brainTransplant(oldC, *this);
         Map[FuncName] = newC;
         RetVal = true;
-      } else
+      } else {
         // if the current FV constraint is not a definition?
         // then merge.
-        oldC->mergeDeclaration(newC, *this);
+        std::string BaseFailed = "";
+        std::string ReasonFailed = "";
+        oldC->mergeDeclaration(newC, *this, ReasonFailed);
+        bool MergingFailed = BaseFailed != ReasonFailed;
+        if (MergingFailed) {
+          clang::DiagnosticsEngine &DE = C->getDiagnostics();
+          unsigned MergeFailID = DE.getCustomDiagID(DiagnosticsEngine::Fatal,
+                                           "merging failed for %q0 due to %1");
+          const auto Pointer = reinterpret_cast<intptr_t>(FD);
+          const auto Kind = clang::DiagnosticsEngine::ArgumentKind::ak_nameddecl;
+          auto DiagBuilder = DE.Report(FD->getLocation(), MergeFailID);
+          DiagBuilder.AddTaggedVal(Pointer, Kind);
+          DiagBuilder.AddString(ReasonFailed);
+        }
+        if(MergingFailed) {
```

Comment:
I'm confused about this. Does just doing `AddString` to `DiagBuilder` above cause the message to be printed? If not, then this conditional will eat the message by causing the process to die, right?

Reply:
It doesn't by itself. The way this works is as you observed. When the Diagnostic Builder goes out of scope, the associated error is output! The reason why I have a separate (but identical) conditional to exit is that since the diagnostic only fires when it goes out of scope, if I have the call to `exit` in the same block, the program will exit like expected, but no diagnostics will be provided. A bit hacky, but it's the only thing that seems to work if we want to immediately exit and not "soldier through."

Reply:
I don't think we can just exit. This is in library code, and may be part of `clangd`. There needs to be some other way to abort execution. I wonder if there are other clangtools that might provide a model for what to do...

Reply:
Mind clarifying what's wrong with `exit()`? I did a quick Google search for standard methods of termination with clang and llvm as keywords, and it seems `std::terminate()` is a function that's used internally that might be of use. But I don't think the intent of your comment was "replace the call to exit with the call to something else that does the same thing", but rather, that exiting abruptly might break something? If so, do you know what it breaks? I'm not too familiar with how `clangd` could complicate things.

Reply:
`clangd` is a server ("clang daemon" == server running all the time). So the point is that if just do `exit` in the middle of our library, linked to `clangd`, then the `clangd` abruptly terminates. Instead, we just want to terminate our analysis, not terminate the whole program that called into it.

Reply:
This might mean we have to do a little restructuring of the tool itself. Maybe @jackastner or @mattmccutchen-cci have ideas about how this could be done.

Reply:
Without being familiar with the code (and not wanting to drop everything to try to become familiar with it right now): I don't know what the call stack for our analysis looks like, and it seems it might be a lot of work to unwind it safely and return to `clangd` except perhaps by throwing and catching a C++ exception (if that's acceptable in our code; I don't see any use of C++ exceptions so far).  In any case, if 3C's `clangd` mode maintains any data structures incrementally across analysis attempts, we need to make sure the data structures can recover from termination of the analysis (due to inconsistent merged declarations or any other cause of termination we may add in the future).

If that isn't enough of a hint, I guess you have to wait for @jackastner.

Reply:
I'm not sure how easy/hard it is, but you might look at the structure of the Checked C typechecker, which is I think in `clang/lib/Sema/*.cpp` (at least in part); it extends the existing C typechecker. That code must have the same issue: If things go south they want to signal an error, but they also want to exit gracefully. Maybe they use exceptions to do this, not sure.